src/Microsoft.ServiceFabric.Actors.Migration/AmbiguousActorIdHandler.cs

```diff
@@ -0,0 +1,260 @@
+        }
+
+        /// <summary>
+        /// Populates the Ambigious Actor Id cache on failover
+        /// </summary>
+        /// <param name="cancellationToken">Cancellation Token</param>
+        /// <returns>Task</returns>
+        public async Task PopulateCache(CancellationToken cancellationToken)
+        {
+            await this.InitializeMetadataTableAsync(cancellationToken);
+
+            ConditionalValue<HashSet<string>> metadataTableValue;
+            using (var tx = this.stateProvider.GetStateManager().CreateTransaction())
+            {
+                metadataTableValue = await this.metadataTable.TryGetValueAsync(tx, MetaDataKey, TimeSpan.FromSeconds(60), cancellationToken);
+            }
+
+            if (metadataTableValue.HasValue)
+            {
+                this.metadataCache = metadataTableValue.Value;
+            }
+        }
+
+        internal void AddRange(List<string> stringActorIds)
+        {
+            stringActorIds.ForEach(x => this.Add(x));
+        }
+
+        internal async void Add(string stringActorId)
+        {
+            if (stringActorId.StartsWith(MigrationConstants.StringTypeActorIdPrefix))
+            {
+                stringActorId = stringActorId.Substring(MigrationConstants.StringTypeActorIdPrefix.Length, stringActorId.Length - MigrationConstants.StringTypeActorIdPrefix.Length - 1);
+            }
+
+            if (!this.IsAmbiguousActorId(stringActorId))
+            {
+                this.metadataCache.Add(stringActorId);
+                await this.InitializeMetadataTableAsync(CancellationToken.None);
+                using (var tx = this.stateProvider.GetStateManager().CreateTransaction())
+                {
+                    await this.metadataTable.AddOrUpdateAsync(tx, MetaDataKey, this.metadataCache, (k, v) => this.metadataCache);
+                    await tx.CommitAsync();
+                }
+            }
+            else
+            {
+                this.ReportPartitionHealthAndThrow();
+            }
+        }
+
+        internal bool IsAmbiguousActorId(string stringActorId)
+        {
+            var count = 0;
+            var tempStr = stringActorId;
+
+            // drop last token. which could be statename or reminder name
+            if (stringActorId.Contains("_"))
+            {
+                var lastIndexOfUnderscore = stringActorId.LastIndexOf('_');
+                tempStr = stringActorId.Substring(0, lastIndexOfUnderscore);
+            }
+
+            for (int iterations = 0; iterations < stringActorId.Split('_').Count(); iterations++)
+            {
+                count += this.metadataCache.Select(n => n.Contains(tempStr)).Count();
+
+                var lastIndexOfUnderscore = tempStr.LastIndexOf('_');
+                tempStr = tempStr.Substring(0, lastIndexOfUnderscore);
+            }
+
+            if (count > 1)
+            {
+                foreach (Type resolver in this.resolvers)
```

Comment:
Resolver are used while fetching the actual state/reminders. Cache population is with presence keys.